nni/algorithms/compression/pytorch/pruning/iterative_pruner.py

```diff
@@ -0,0 +1,584 @@
+from nni.compression.pytorch.compressor import Pruner
+from .constants import MASKER_DICT, MAX_EPOCHS
+from .dependency_aware_pruner import DependencyAwarePruner
+
+__all__ = ['AGPPruner', 'ADMMPruner', 'SlimPruner', 'TaylorFOWeightFilterPruner', 'ActivationAPoZRankFilterPruner', 'ActivationMeanRankFilterPruner', ]
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+class IterativePruner(DependencyAwarePruner):
+    """
+    Prune model during the training process. 
+    """
+
+    def __init__(self, model, config_list, optimizer=None, pruning_algorithm='level', trainer=None, criterion=None, training_epochs=1, dependency_aware=False, dummy_input=None, **algo_kwargs):
+        """
+        Parameters
+        ----------
+        model : torch.nn.Module
+            Model to be pruned
+        config_list : list
+            List on pruning configs
+        pruning_algorithm: str
+            algorithms being used to prune model
+        optimizer: torch.optim.Optimizer
+            Optimizer used to train model
+        trainer: function
+            Function used to train the model.
+            Users should write this function as a normal function to train the Pytorch model
+            and include `model, optimizer, criterion, epoch, callback` as function arguments.
+        criterion: function
+            Function used to calculate the loss between the target and the output.
+        training_epochs : int
+            Totoal number of epochs for training.
+        dependency_aware: bool
+            If prune the model in a dependency-aware way.
+        dummy_input : torch.Tensor
+            The dummy input to analyze the topology constraints. Note that, the dummy_input should on the same device with the model.
+        algo_kwargs: dict
+            Additional parameters passed to pruning algorithm masker class
+        """
+        super().__init__(model, config_list, optimizer, pruning_algorithm, dependency_aware, dummy_input, **algo_kwargs)
+
+        self.training_epochs = training_epochs
+        self._trainer = trainer
+        self._criterion = criterion
+
+    def compress(self):
+        training = self.bound_model.training
+        self.bound_model.train()
+        for epoch in range(MAX_EPOCHS):
+            self._trainer(self.bound_model, optimizer=self.optimizer,
+            criterion=self._criterion, epoch=epoch, callback=self._callback)
+            if epoch >= self.training_epochs:
+                break
+        self.bound_model.train(training)
+        self._get_threshold()
+
+        self.update_mask()
+        self.patch_optimizer(self.update_mask)
+        
+        return self.bound_model
+
+
+    def _get_threshold(self):
+        """
+        Function to get global threshold
+        """
+        pass
+
+    def _callback(self):
+        """
+        Callback function to do additonal optimization
+        """
+        pass
+
+    def calc_mask(self, wrapper, wrapper_idx=None):
```

Comment:
seems the implementation is the same in `DependencyAwarePruner`

Reply:
Thanks for point out that. Remove the duplicate functions.